### ISSUE TITLE
mc mirror or cp (or some other command that need at least one worker) crash on ARMv7

### ISSUE THREAD

**Issue by protheusfr** [minio/mc#2387]
## Expected behaviour
Everything works :)
## Actual behaviour
```
mc mirror /mnt/test m1/test
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x12838]

goroutine 40 [running]:
sync/atomic.loadUint64(0x10faf46c, 0x346030, 0x10)
	/usr/local/go/src/sync/atomic/64bit_arm.go:10 +0x3c
github.com/minio/mc/cmd.(*ParallelManager).addWorker(0x10faf440)
	/home/pi/go/src/github.com/minio/mc/cmd/parallel-manager.go:62 +0x28
github.com/minio/mc/cmd.newParallelManager(0x10e6f440, 0x75591330, 0x10ffdee0, 0x75591330, 0x10ffdee0)
	/home/pi/go/src/github.com/minio/mc/cmd/parallel-manager.go:142 +0x108
github.com/minio/mc/cmd.(*mirrorJob).startMirror(0x10fff680, 0x633ce0, 0x11006340, 0x11004010, 0x10e6f580)
	/home/pi/go/src/github.com/minio/mc/cmd/mirror-main.go:449 +0x58
created by github.com/minio/mc/cmd.(*mirrorJob).mirror
	/home/pi/go/src/github.com/minio/mc/cmd/mirror-main.go:527 +0xa8
```
## Steps to reproduce the behaviour
go get and compile mc on a raspberry 3 under Debian Stretch.

## mc version
- Version: 2018-02-10T10:13:19Z
Release-tag: DEVELOPMENT.2018-02-10T10-13-19Z
Commit-id: 3987f1405aa7a0faaff019cc3b55da4ded9ebbe5

## System information
go version go1.9.4 linux/arm

This is caused by `atomic.LoadUint64` on 32 bits systems.
Go documentation say : 
```
On both ARM and x86-32, it is the caller's responsibility to arrange
for 64-bit alignment of 64-bit words accessed atomically. The first word
in a variable or in an allocated struct, array, or slice can be relied
upon to be 64-bit aligned.
```

**Comment by protheusfr** [comment]
Suggested fix : 

```
Index: cmd/parallel-manager.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- cmd/parallel-manager.go	(revision 2d221ba8da900731e70e08ac4e11687370a54ff0)
+++ cmd/parallel-manager.go	(revision 451a2457177e92784d3e5cc930bd2b2407a5aad3)
@@ -20,6 +20,7 @@
 	"sync"
 	"sync/atomic"
 	"time"
+	"unsafe"
 )
 
 const (
@@ -46,7 +47,7 @@
 	bandwidthPrev float64
 
 	// Current threads number
-	threadsNum uint64
+	threadsNum [12]byte
 
 	// Channel to receive tasks to run
 	queueCh chan func() URLs
@@ -56,17 +57,24 @@
 	stopMonitorCh chan struct{}
 }
 
+func (p *ParallelManager) threadsNumValue64BitAlignedPtr() *uint64 {
+	if uintptr(unsafe.Pointer(&p.threadsNum))%8 == 0 {
+		return (*uint64)(unsafe.Pointer(&p.threadsNum))
+	}
+	return (*uint64)(unsafe.Pointer(&p.threadsNum[4]))
+}
+
 // addWorker creates a new worker to process tasks
 func (p *ParallelManager) addWorker() {
 
-	if atomic.LoadUint64(&p.threadsNum) >= maxParallelWorkers {
+	if atomic.LoadUint64(p.threadsNumValue64BitAlignedPtr()) >= maxParallelWorkers {
 		// Number of maximum workers is reached, no need to
 		// to create a new one.
 		return
 	}
 
 	// Update number of threads
-	atomic.AddUint64(&p.threadsNum, 1)
+	atomic.AddUint64(p.threadsNumValue64BitAlignedPtr(), 1)
 
 	// Start a new worker
 	p.wg.Add(1)
@@ -132,7 +140,7 @@
 	p := &ParallelManager{
 		pb:            pb,
 		wg:            &sync.WaitGroup{},
-		threadsNum:    0,
+		threadsNum:    [12]byte{},
 		stopMonitorCh: make(chan struct{}),
 		queueCh:       make(chan func() URLs),
 		resultCh:      resultCh,
```

**Comment by harshavardhana** [comment]
We do not test on arm platforms. Just that source compilation is supported so not surprised if we have these issues.

Also the suggested fix is a bit involved with endianess. I would simply use atomic 32 bit versions instead.

**Comment by protheusfr** [comment]
Indeed I suspect 2^32 workers would be enough.

Bastien Vigneron

**Comment by harshavardhana** [comment]
Fixed by #2389

**Comment by kannappanr** [comment]
Closing the issue as it is fixed by #2389
